{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Display gather results w/taxonomy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from charcoal import utils\n",
    "import sourmash\n",
    "from sourmash.lca import taxlist, LineagePair\n",
    "import collections\n",
    "\n",
    "import plotly.graph_objects as go\n",
    "\n",
    "import sourmash_sankey\n",
    "import importlib\n",
    "\n",
    "import os.path\n",
    "\n",
    "import screed\n",
    "\n",
    "from sourmash.lca.command_index import load_taxonomy_assignments\n",
    "from sourmash.lca import LCA_Database\n",
    "\n",
    "from charcoal.lineage_db import LineageDB\n",
    "from charcoal.utils import (gather_at_rank, get_ident, ContigGatherInfo)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def do(lineages_csv, genome_filename, genome_sig_filename, matches_filename):\n",
    "    genomebase = os.path.basename(genome_filename)\n",
    "    match_rank = 'genus'\n",
    "\n",
    "    # load taxonomy CSV                                                         \n",
    "    tax_assign, _ = load_taxonomy_assignments(lineages_csv,\n",
    "                                              start_column=3)\n",
    "    print(f'loaded {len(tax_assign)} tax assignments.')\n",
    "\n",
    "    # load the genome signature                                                 \n",
    "    genome_sig = sourmash.load_one_signature(genome_sig_filename)\n",
    "\n",
    "    # load all of the matches from search --containment in the database         \n",
    "    with open(matches_filename, 'rt') as fp:\n",
    "        try:\n",
    "            siglist = list(sourmash.load_signatures(fp, do_raise=True,\n",
    "                                                    quiet=False))\n",
    "        except sourmash.exceptions.SourmashError:\n",
    "            siglist = []\n",
    "    print(f\"loaded {len(siglist)} matches from '{matches_filename}'\")\n",
    "\n",
    "    # Hack for examining members of our search database: remove exact matches.  \n",
    "    new_siglist = []\n",
    "    for ss in siglist:\n",
    "        if genome_sig.similarity(ss) == 1.0:\n",
    "            print(f'removing an identical match: {ss.name()}')\n",
    "        else:\n",
    "            new_siglist.append(ss)\n",
    "    siglist = new_siglist\n",
    "\n",
    "    # if, after removing exact match(es), there is nothing left, quit.                                                                 \n",
    "    if not siglist:\n",
    "        print('no non-identical matches for this genome, exiting.')\n",
    "        return {}\n",
    "\n",
    "    # construct a template minhash object that we can use to create new 'uns    \n",
    "    empty_mh = siglist[0].minhash.copy_and_clear()\n",
    "    ksize = empty_mh.ksize\n",
    "    scaled = empty_mh.scaled\n",
    "    moltype = empty_mh.moltype\n",
    "\n",
    "    # create empty LCA database to populate...                                  \n",
    "    lca_db = LCA_Database(ksize=ksize, scaled=scaled, moltype=moltype)\n",
    "    lin_db = LineageDB()\n",
    "\n",
    "    # ...with specific matches.                                                 \n",
    "    for ss in siglist:\n",
    "        ident = get_ident(ss)\n",
    "        lineage = tax_assign[ident]\n",
    "\n",
    "        lca_db.insert(ss, ident=ident)\n",
    "        lin_db.insert(ident, lineage)\n",
    "\n",
    "    print(f'loaded {len(siglist)} signatures & created LCA Database')\n",
    "\n",
    "    print('')\n",
    "    print(f'reading contigs from {genomebase}')\n",
    "\n",
    "    screed_iter = screed.open(genome_filename)\n",
    "    contigs_tax = {}\n",
    "    for n, record in enumerate(screed_iter):\n",
    "       # look at each contig individually                                      \n",
    "        mh = empty_mh.copy_and_clear()\n",
    "        mh.add_sequence(record.sequence, force=True)\n",
    "\n",
    "        # collect all the gather results at genus level, together w/counts;     \n",
    "        # here, results is a list of (lineage, count) tuples.                   \n",
    "        results = list(gather_at_rank(mh, lca_db, lin_db, match_rank))\n",
    "\n",
    "        # store together with size of sequence.                                 \n",
    "        info = ContigGatherInfo(len(record.sequence), len(mh), results)\n",
    "        contigs_tax[record.name] = info\n",
    "\n",
    "    print(f\"Processed {len(contigs_tax)} contigs.\")\n",
    "    return contigs_tax\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "importlib.reload(sourmash_sankey)\n",
    "from sourmash_sankey import GenomeSankeyFlow\n",
    "\n",
    "def make_sankey_fig(title, genome_lineage, contigs_info):\n",
    "    obj = GenomeSankeyFlow()\n",
    "    \n",
    "    counts = collections.Counter()\n",
    "    for contig_name, gather_info in contigs_info.items():\n",
    "        contig_taxlist = gather_info.gather_tax\n",
    "\n",
    "        # iterate over each contig match and summarize counts.              \n",
    "        # note - here we can stop at first one, or track them all.          \n",
    "        # note - b/c gather counts each hash only once, these               \n",
    "        #     are non-overlapping                                           \n",
    "        total_hashcount = 0\n",
    "        for lin, hashcount in contig_taxlist:\n",
    "            counts[lin] += hashcount\n",
    "            total_hashcount += hashcount\n",
    "\n",
    "        # track missing => unassigned lineage\n",
    "        unident = gather_info.num_hashes - total_hashcount\n",
    "        counts[obj.unassigned_lin] += unident\n",
    "    \n",
    "    # set the color of the main lineage\n",
    "    genome_lineage = tuple(genome_lineage)\n",
    "    obj.colors[genome_lineage] = \"lightseagreen\"\n",
    "    \n",
    "    # for phylum level disagreements, let's go with palevioletred\n",
    "    for lin in counts:\n",
    "        if not utils.is_lineage_match(lin, genome_lineage, 'phylum'):\n",
    "            obj.colors[lin] = 'palevioletred'\n",
    "            \n",
    "    # assign unassigned to good lineage, maybe?\n",
    "    counts[genome_lineage] += counts[obj.unassigned_lin]\n",
    "    del counts[obj.unassigned_lin]\n",
    "    \n",
    "    obj.make_links(genome_lineage, counts)\n",
    "    fig = obj.make_plotly_fig(title)\n",
    "    \n",
    "    return fig\n",
    "\n",
    "genome_lin = 'd__Archaea,p__Thermoplasmatota,c__Poseidoniia,o__Poseidoniales,f__Thalassoarchaeaceae,g__MGIIb-O3,s__MGIIb-O3 sp002722735'\n",
    "genome_lin = utils.make_lineage(genome_lin)\n",
    "\n",
    "lineages_csv = '../charcoal/db/gtdb-release89-lineages.csv'\n",
    "\n",
    "contigs_info = do(lineages_csv, 'tobg_np-110.fa', 'tobg_np-110.k31.sig', 'tobg_np-110.sig.matches')\n",
    "make_sankey_fig('tobg_np-110', genome_lin, contigs_info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (myenv)",
   "language": "python",
   "name": "myenv"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
